dali/pipeline/operator/sequence_op.h

```diff
@@ -0,0 +1,471 @@
+      }
+    }
+  }
+
+  virtual void CoalesceOutputShapes(std::vector<OutputDesc> &output_desc,
+                                    const workspace_t<Backend> &ws) {
+    for (size_t output_idx = 0; output_idx < output_desc.size(); output_idx++) {
+      const auto &expand_desc = GetOutputExpandDesc(ws, output_idx);
+      const auto &shape = output_desc[output_idx].shape;
+      DALI_ENFORCE(
+          shape.num_samples() == expand_desc.NumElements(),
+          make_string("Unexpected number of samples or frames inferred in the operator for output ",
+                      output_idx, ". Expected ", expand_desc.NumElements(),
+                      " but the operator returned ", shape.num_samples(), "."));
+      output_desc[output_idx].shape = fold_outermost_like(shape, expand_desc.ExpandedShape());
+    }
+  }
+
+  virtual bool ProcessOutputDesc(std::vector<OutputDesc> &output_desc,
+                                 const workspace_t<Backend> &ws, bool is_inferred) {
+    if (is_inferred && expand_) {
+      CoalesceOutputShapes(output_desc, ws);
+    }
+    return is_inferred;
+  }
+
+  virtual void ExpandArgment(const workspace_t<Backend> &ws, const std::string &arg_name,
+                             const TensorVector<CPUBackend> &arg_tensor) {
+    const auto &expand_desc = GetArgExpandDesc(arg_name);
+    auto expanded_arg = ExpandArgumentLike(arg_tensor, arg_name, expand_desc);
+    auto expanded_handle = std::make_shared<TensorVector<CPUBackend>>(std::move(expanded_arg));
+    AddArgumentInputHelper(arg_name, expanded_handle, {ExpandDescFrameInfoFn{&expand_desc}});
+  }
+
+  virtual void ExpandArguments(const workspace_t<Backend> &ws) {
+    for (const auto &arg_input : ws) {
+      auto shared_tvec = arg_input.second.tvec;
+      assert(shared_tvec);
+      ExpandArgment(ws, arg_input.first, *shared_tvec);
+    }
+  }
+
+  virtual const ExpandDesc &GetOutputExpandDesc(const workspace_t<Backend> &ws, int output_idx) {
+    (void)ws;
+    return GetInputExpandDesc(output_idx);
+  }
+
+  virtual const ExpandDesc &GetArgExpandDesc(const std::string &arg_name) {
+    (void)arg_name;
+    return GetInputExpandDesc(0);
+  }
+
+  bool HasPerFrameArgInput(const workspace_t<Backend> &ws) {
+    for (const auto &arg_input : ws) {
+      auto shared_tvec = arg_input.second.tvec;
+      assert(shared_tvec);
+      if (IsPerFrame(*shared_tvec)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  bool IsPerFrame(const TensorVector<CPUBackend> &arg_tensor) {
+    const auto &layout = arg_tensor.GetLayout();
+    return layout.size() > 0 && layout[0] == 'F';
+  }
+
+  template <typename InputBackend>
+  ws_input_t<InputBackend> ExpandInputHelper(const workspace_t<Backend> &ws, int input_idx,
+                                             int num_expand_dims) {
+    static_assert(is_shared_ptr<ws_input_t<InputBackend>>::value,
+                  "Workspace input handle expected to be shared_ptr");
```

Comment:
This seems overly defensive - after all, there's not much choice of what `workspace_t<InputBackend>` can be.

Reply:
Sure.